test/compilable/test20138.d

```diff
@@ -0,0 +1,17 @@
+alias C = const int;
+static assert(is(shared(C) U == shared U) && is(U == C));
+static assert(is(shared(C) == shared U, U) && is(U == C));
+
+
+alias I = inout int;
+static assert(is(shared(I) U == shared U) && is(U == I));
+static assert(is(shared(I) == shared U, U) && is(U == I));
+
+alias IC = inout const int;
+static assert(is(shared(IC) U == shared U) && is(U == IC));
+static assert(is(shared(IC) == shared U, U) && is(U == IC));
+
+alias S = shared int;
+static assert(is(const(S) U == const U) && is(U == shared int));
+static assert(is(inout(S) U == inout U) && is(U == shared int));
+static assert(is(inout(const S) U == inout(const U)) && is(U == shared int));
```

Comment:
Suggest adding `static assert(is(inout const S U == inout U) && is(U == const shared S));`

Reply:
> Suggest adding `static assert(is(inout const S U == inout U) && is(U == const shared S));`

I'd rather leave that out for now, because it's less obvious how to handle it.

For the others, I just changed `MATCH.constant` to `MATCH.exact`. For this combination, `deduceTypeHelper` currently returns `MATCH.nomatch`. But it works in IFTI, so there must be something else going on. My guess is that it's handled by `deduceTypeHelper`, which isn't even called for `is` expressions.

Reply:
> My guess is that it's handled by `deduceTypeHelper`, which isn't even called for `is` expressions.

That was supposed to be: My guess is that it's handled by **`deduceWildHelper`**

Reply:
I was gonna say we could do something like this:
```
    *at = t.mutableOf().unSharedOf();
    if ((t.mod & MODFlags.const_) && !(tParam.mod & MODFlags.const_))
        *at = (*at).constOf();
    if ((t.mod & MODFlags.shared_) && !(tParam.mod & MODFlags.shared_))
        *at = (*at).sharedOf();
    if ((t.mod & MODFlags.immutable_) && !(tParam.mod & MODFlags.immutable_))
        *at = (*at).immutableOf();
    if ((t.mod & MODFlags.wild) && !(tParam.mod & MODFlags.wild))
        *at = (*at).wildOf();
```
However, apparently calling `sharedOf()` on a `const` type doesn't return a `const shared` type, and likewise for the other `*Of()`s. This somewhat disappoints me.